src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebHostEnvironmentTests.cs

```diff
@@ -58,16 +60,94 @@ public void ApplyEnvironmentSettingsUsesTheCorrectKeysAndProperties()
+            environment.WebRootPath = tempPath;
+
+            Assert.Equal(tempPath, environment.WebRootPath);
+            Assert.Equal(tempPath, environment.ContentRootPath);
+
+            Assert.IsType<PhysicalFileProvider>(environment.ContentRootFileProvider);
+            Assert.IsType<PhysicalFileProvider>(environment.WebRootFileProvider);
+
+            Assert.Equal(tempPath, ((PhysicalFileProvider)environment.ContentRootFileProvider).Root);
+            Assert.Equal(tempPath, ((PhysicalFileProvider)environment.WebRootFileProvider).Root);
+        }
+
+        [Fact]
+        public void RelativePathsAreMappedToFullPaths()
+        {
+            var environment = new WebHostEnvironment();
+            var relativeRootPath = "/some-relative-path";
+            var relativeSubPath = "/some-other-relative-path";
+            
+            // ContentRootPath is mapped relative to AppContext.BaseDirectory
+            environment.ContentRootPath = relativeRootPath;
+            Assert.Equal(Path.Combine(AppContext.BaseDirectory, relativeRootPath), environment.ContentRootPath);
+
+            // WebRootPath is mapped relative to ContentRootPath
+            environment.WebRootPath = relativeSubPath;
+            Assert.Equal(Path.Combine(AppContext.BaseDirectory, relativeRootPath, relativeSubPath), environment.WebRootPath);
+        }
+
+        [Fact]
+        public void CannotUnsetPaths()
+        {
+            var environment = new WebHostEnvironment();
+            var webRootPath = Path.GetTempPath();
+
+            environment.WebRootPath = webRootPath;
+
+            Assert.Equal(webRootPath, environment.WebRootPath);
+            Assert.Equal(webRootPath, ((PhysicalFileProvider)environment.WebRootFileProvider).Root);
+
+            // Setting WebRootPath to null no-ops
+            environment.WebRootPath = null;
+            Assert.Equal(webRootPath, environment.WebRootPath);
+            Assert.Equal(webRootPath, ((PhysicalFileProvider)environment.WebRootFileProvider).Root);
+
+            // Setting ContentRootPath to null falls back to current directory
+            environment.ContentRootPath = null;
+            Assert.Equal(AppContext.BaseDirectory, environment.ContentRootPath);
+            Assert.Equal(AppContext.BaseDirectory, ((PhysicalFileProvider)environment.ContentRootFileProvider).Root);
+        }
+
+        [Fact]
+        public void SetContentRootAfterRelativeWebRoot()
+        {
+            var environment = new WebHostEnvironment();
+            var webRootPath = "/some-relative-path";
+            var tempPath = Path.GetTempPath();
+
+            environment.WebRootPath = webRootPath;
+
+            Assert.Equal(Path.Combine(AppContext.BaseDirectory, webRootPath), environment.WebRootPath);
+
+            // Setting the ContentRootPath after setting a relative WebRootPath
+            // does not modify the WebRootPath
+            environment.ContentRootPath = tempPath;
```

Comment:
I changed my mind about this. I think we do need to preserve the relative web root path and re-ResolvePathToRoot when the ContentRootPath changes in order to keep the behavior consistent with the generic host when you call UseWebRoot() before UseContentRoot().